### PR TITLE
Add the Perplexity metric

### PR DESCRIPTION
**Pull request by bryant1410**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/references/functional.rst [hunk 1]
```diff
@@ -525,6 +525,12 @@ match_error_rate [func]
 .. autofunction:: torchmetrics.functional.match_error_rate
     :noindex:
 
+perplexity [func]
+~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (Borda) [inline docs/source/references/functional.rst:529]:
```suggestion
~~~~~~~~~~~~~~~~~
```

File: tests/text/test_perplexity.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
```
Comment (Borda) [inline tests/text/test_perplexity.py:1]: lets use the testing template, cc @stancld

File: torchmetrics/functional/text/perplexity.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+# limitations under the License.
+
+from typing import Optional, Tuple
+
+from torch import Tensor, isnan
+
+
+# From https://github.com/pytorch/pytorch/issues/21987#issuecomment-539402619
+# From PyTorch v1.10, this is officially supported.
+def nanmean(v: Tensor, *args, inplace: bool = False, **kwargs) -> Tensor:  # type: ignore
```
Comment (Borda) [inline torchmetrics/functional/text/perplexity.py:22]: rather protected or move to general functions

**Review by Borda — CHANGES_REQUESTED** [review]
need requires testing :rabbit:

File: torchmetrics/functional/text/perplexity.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    Examples:
+        >>> from torch import tensor
+        >>> probs = tensor([[.2, .04, .8], [.34, .12, .56]])
+        >>> mask = tensor([[True, True, False], [True, True, True]])
+        >>> perplexity(probs, mask)
+        tensor(7.3522)
+    """
+    total, count = _perplexity_update(probs, mask)
+    return _perplexity_compute(total, count)
```
Comment (stancld) [inline torchmetrics/functional/text/perplexity.py:81]: I would suggest using `preds, target` inputs instead of `probs` to be consistent with other metrics in TM. Also, plenty of models (e.g. all from `transformers`) return a vector of logits for a given tokens, therefore, it would be more convenient for our users if they pass these logits alongside with target (labels) into our metric.

It can be also considered `target` to be an `Optional` argument when the format of `preds` mirrors your current use case.

**Review by stancld — COMMENTED** [review]
@bryant1410 Thank you a lot for contributing! 🎉 

I have one issue regarding the API of this new metric so that it will better correspond to the design of `torchmetrics`.

File: tests/text/test_perplexity.py (hunk 2 above)
Comment (stancld) [reply tests/text/test_perplexity.py:1]: @bryant1410 Could you please write tests according you our template? In case we have no direct reference package and function to use, one good example can be found [here (EED tests)](https://github.com/PyTorchLightning/metrics/blob/master/tests/text/test_eed.py).

Definitely let me know should you need any help! :]

**Comment by bryant1410** [comment]
> need requires testing rabbit

@Borda what do you mean? To add more tests? Specifically, testing what behavior?

**Comment by Borda** [comment]
> what do you mean? To add more tests? Specifically, testing what behavior?

yes, use the testing template which is made for complete metrics testing, see for example:
https://github.com/PyTorchLightning/metrics/blob/master/tests/text/test_cer.py

**Comment by bryant1410** [comment]
> yes, use the testing template which is made for complete metrics testing, see for example:
> https://github.com/PyTorchLightning/metrics/blob/master/tests/text/test_cer.py

I see. I can do that. The only issue I see is that I don't have a reference implementation, so the metric function to compare against would be hardcoded.

**Comment by Borda** [comment]
> The only issue I see is that I don't have a reference implementation, so the metric function to compare against would be hardcoded.

@stancld could you pls help here? :)

**Comment by stancld** [comment]
@bryant1410 Any updates on here regarding [this](https://github.com/PyTorchLightning/metrics/pull/922#discussion_r839785057)? :]

**Comment by bryant1410** [comment]
> @bryant1410 Any updates on here regarding [this](https://github.com/PyTorchLightning/metrics/pull/922#discussion_r839785057)? :]

Hey, I've been busy. I don't have an ETA. But if you want to continue it, please feel free!

**Comment by Borda** [comment]
@bryant1410 @stancld how is it going here? :rabbit: 
we are considering a new feature release next week, do you think we can finish this addition by then so it will be included? :pray:

**Comment by bryant1410** [comment]
@Borda i can't finish it this week. I'm happy to let somebody continue it if they want. I can otherwise continue it later

**Comment by stancld** [comment]
@Borda Looks like `InfoLM` is now really ready for review, then I hope I'll prepare this one tomorrow O:]

File: src/torchmetrics/functional/text/perplexity.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        raise ValueError(
+            "Input tensors `preds` and `target` are expected to have equaling first two dimensions, "
+            f"[batch_size, seq_len], but got {preds.shape[:2]} and {target.shape}."
+        )
+    if preds.dtype not in _TORCH_FLOAT_OR_DOUBLE:
+        raise TypeError(
+            f"Input tensor `preds` is expected to be of a type one of {_TORCH_FLOAT_OR_DOUBLE} but got {preds.dtype}."
+        )
+    if target.dtype != torch.int64:
+        raise TypeError(f"Input tensor `target` is expected to be of a type {torch.int64} but got {target.dtype}.")
```
Comment (Borda) [inline src/torchmetrics/functional/text/perplexity.py:65]:
```suggestion
        raise ValueError(
            "Input tensor `preds` is expected to have 3 dimensions, [batch_size, seq_len, vocab_size],"
            f" but got {len(preds.shape)}."
        )
    if len(target.shape) != 2:
        raise ValueError(
            "Input tensor `target` is expected to have 2 dimensions, [batch_size, seq_len],"
            f" but got {len(target.shape)}."
        )
    if preds.shape[:2] != target.shape:
        raise ValueError(
            "Input tensors `preds` and `target` are expected to have equaling first two dimensions,"
            f" [batch_size, seq_len], but got {preds.shape[:2]} and {target.shape}."
        )
    if preds.dtype not in _TORCH_FLOAT_OR_DOUBLE:
        raise TypeError(
            f"Input tensor `preds` is expected to be of a type one of {_TORCH_FLOAT_OR_DOUBLE} but got {preds.dtype}."
        )
    if target.dtype != torch.int64:
        raise TypeError(f"Input tensor `target` is expected to be of a type {torch.int64} but got {target.dtype}.")
```

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/functional/text/perplexity.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        raise ValueError(
+            "Input tensor `target` is expected to have 2 dimensions, [batch_size, seq_len],"
+            f" but got {len(target.shape)}."
+        )
+    if preds.shape[:2] != target.shape:
+        raise ValueError(
+            "Input tensors `preds` and `target` are expected to have equaling first two dimensions,"
+            f" [batch_size, seq_len], but got {preds.shape[:2]} and {target.shape}."
+        )
+    if preds.dtype not in _TORCH_FLOAT_OR_DOUBLE:
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/text/perplexity.py:60]: reason why `half` is not supported?

File: src/torchmetrics/text/perplexity.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    Args:
+        ignore_index:
+            Integer specifying a target class to ignore. If given, this class index does not contribute
+            to the returned score.
+        kwargs:
+            Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+
+    Returns:
+        Perplexity
```
Comment (SkafteNicki) [inline src/torchmetrics/text/perplexity.py:36]: It fine to just have this in the compute method
```suggestion
```

File: src/torchmetrics/text/perplexity.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        >>> preds = torch.rand(2, 8, 5, generator=torch.manual_seed(22))
+        >>> target = torch.randint(5, (2, 8), generator=torch.manual_seed(22))
+        >>> target[0, 6:] = -100
+        >>> metric = Perplexity(ignore_index=-100)
+        >>> metric(preds, target)
+        tensor(5.2545)
+    """
+    is_differentiable = True
+    higher_is_better = False
+    total_log_probs: Tensor
```
Comment (SkafteNicki) [inline src/torchmetrics/text/perplexity.py:49]:
```suggestion
    full_state_update = False
    total_log_probs: Tensor
```

File: src/torchmetrics/text/perplexity.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    count: Tensor
+
+    def __init__(
+        self,
+        ignore_index: Optional[int] = None,
+        **kwargs: Dict[str, Any],
+    ):
+        super().__init__(compute_on_step=None, **kwargs)
```
Comment (SkafteNicki) [inline src/torchmetrics/text/perplexity.py:57]:
```suggestion
        super().__init__(**kwargs)
```

File: src/torchmetrics/text/perplexity.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        self.ignore_index = ignore_index
```
Comment (SkafteNicki) [inline src/torchmetrics/text/perplexity.py:58]:
```suggestion
        if ignore_index is not None or not isinstance(ignore_index, int):
            raise ValueError(f"Argument `ignore_index` expected to either be `None` or an `int` but got {ignore_index}")
        self.ignore_index = ignore_index
```

File: tests/unittests/text/test_perplexity.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+from unittests.text.inputs import (
+    MASK_INDEX,
+    _logits_inputs_fp32,
+    _logits_inputs_fp32_with_mask,
+    _logits_inputs_fp64,
+    _logits_inputs_fp64_with_mask,
+)
+
+
+def _sk_perplexity(preds, target, ignore_index):
```
Comment (SkafteNicki) [inline tests/unittests/text/test_perplexity.py:32]: the sk usually refers to sklearn so lets change it here
```suggestion
def _reference_perplexity(preds, target, ignore_index):
```

File: tests/unittests/text/test_perplexity.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            dist_sync_on_step=dist_sync_on_step,
+            metric_args={"ignore_index": ignore_index},
+        )
+
+    def test_perplexity_fn(self, preds, target, ignore_index):
+        self.run_functional_metric_test(
+            preds,
+            target,
+            metric_functional=perplexity,
+            sk_metric=partial(_sk_perplexity, ignore_index=ignore_index),
```
Comment (SkafteNicki) [inline tests/unittests/text/test_perplexity.py:68]:
```suggestion
            sk_metric=partial(_reference_perplexity, ignore_index=ignore_index),
            dist_sync_on_step=dist_sync_on_step,
            metric_args={"ignore_index": ignore_index},
        )

    def test_perplexity_fn(self, preds, target, ignore_index):
        self.run_functional_metric_test(
            preds,
            target,
            metric_functional=perplexity,
            sk_metric=partial(_reference_perplexity, ignore_index=ignore_index),
```

**Review by SkafteNicki — APPROVED** [review]
